File: examples/go/pagerduty-trigger/handler/handler.go
package function

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"

	handler "github.com/openfaas/templates-sdk/go-http"
	toml "github.com/pelletier/go-toml"
)

const (
	pagerdutyApiPath = "https://events.pagerduty.com/v2/enqueue"
	pdConfigPath     = "/var/openfaas/secrets/pdconfig"

	// Background colors for debug messages.
	bgcHeader    = "\033[95m"
	bgcOkBlue    = "\033[94m"
	bgcOkGreen   = "\033[92m"
	bgcWarning   = "\033[93m"
	bgcFail      = "\033[91m"
	bgcEndC      = "\033[0m"
	bgcBold      = "\033[1m"
	bgcUnderline = "\033[4m"
)

func readEnvVars(pdc *pdConfig) {
	// Stay with default of false unless true is set.
	if os.Getenv("write_debug") == "true" {
		logDebug(pdc, "warn", "debug has been enabled for this function. Sensitive information could be printed to sysout")
	}

	// Stay with default of false unless true is set.
	if os.Getenv("tls_insecure") == "true" {
		pdc.TlsInsecure = false
		logDebug(pdc, "warn", "connections are not TLS secured")
	}
}

// logDebug sends formatted message to stdOut if debug is enabled.
func logDebug(pdc *pdConfig, level string, msg string) {
	if pdc.Debug {
		switch level {
		case "error":
			log.Printf("\n%sError, %s.%s\n", bgcFail, msg, bgcEndC)
		case "warn":
			log.Printf("\n%sWarning, %s.%s\n", bgcWarning, msg, bgcEndC)
		default:
			log.Printf("\n%s\n", msg)
		}
	}
}

// Handle a function invocation.
func Handle(req handler.Request) (handler.Response, error) {
	ctx := context.Background()

	// Load the Pager Duty config file at every function invocation.
	pdCfg, err := loadPdConfig(pdConfigPath)
	if err != nil {
		wrapErr := fmt.Errorf("loading PagerDuty configs: %w", err)
		log.Println(wrapErr.Error())

		return handler.Response{
			Body:       []byte(wrapErr.Error()),
			StatusCode: http.StatusInternalServerError,
		}, wrapErr
	}

	// Transfer event information data to pagerduty data object.
	pdObj, err := parseEvent(req.Body, pdCfg)
	if err != nil {
		wrapErr := fmt.Errorf("parsing event data: %w", err)
		log.Println(wrapErr.Error())

		return handler.Response{
			Body:       []byte(wrapErr.Error()),
			StatusCode: http.StatusInternalServerError,
		}, wrapErr
	}

	// Send information to pagerduty.
	pdRes, err := pdSendRequest(ctx, pdCfg, pagerdutyApiPath, pdObj)
	if err != nil {
		wrapErr := fmt.Errorf("connecting to PagerDuty API: %w", err)
		log.Println(wrapErr.Error())

		return handler.Response{
			Body:       []byte(wrapErr.Error()),
			StatusCode: http.StatusInternalServerError,
		}, wrapErr
	}

	// Display pagerduty response.
	return handler.Response{
		Body:       pdRes,
		StatusCode: http.StatusOK,
	}, nil
}

func loadPdConfig(path string) (*pdConfig, error) {
	var pdc pdConfig

	readEnvVars(&pdc)

	secret, err := toml.LoadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading pdconfig secret file: %w", err)
	}

	if err := secret.Unmarshal(&pdc); err != nil {
		return nil, fmt.Errorf("unmarshalling pdconfig secret file: %w", err)
	}

	if err := validatePdConf(&pdc); err != nil {
		return nil, err
	}

	return &pdc, nil
}

func validatePdConf(pdc *pdConfig) error {
	if pdc.PagerDuty.RoutingKey == "" {
		return errors.New("PagerDuty routing key cannot be empty")
	}

	if pdc.PagerDuty.EventAction == "" {
		return errors.New("PagerDuty event action cannot be empty")
	}

	return nil
}

// parseEvent saves info from the triggering event into a PagerDuty data object.
func parseEvent(req []byte, pdc *pdConfig) (pagerDutyData, error) {
	var ce cloudEvent
	var pd pagerDutyData

	if err := json.Unmarshal(req, &ce); err != nil {
		return pd, err
	}

	if err := isValidEventResp(ce); err != nil {
		return pd, err
	}

	pd.RoutingKey = pdc.PagerDuty.RoutingKey
	pd.EventAction = pdc.PagerDuty.EventAction
	pd.Client = "VMware Event Broker Appliance"
	pd.ClientURL = ce.Source
	pd.Payload.Summary = ce.Data.FullFormattedMessage
	pd.Payload.Timestamp = ce.Data.CreatedTime
	pd.Payload.Source = ce.Source
	pd.Payload.Severity = "info"
	pd.Payload.Component = ce.Data.Vm.Name
	pd.Payload.Group = ce.Data.Host.Name
	pd.Payload.Class = ce.Subject
	pd.Payload.CustomDetails.User = ce.Data.UserName
	pd.Payload.CustomDetails.Datacenter = ce.Data.Datacenter
	pd.Payload.CustomDetails.ComputeResource = ce.Data.ComputeResource
	pd.Payload.CustomDetails.Host = ce.Data.Host
	pd.Payload.CustomDetails.VM = ce.Data.Vm

	return pd, nil
}

func isValidEventResp(event cloudEvent) error {
	if event.Data.Vm == nil || event.Data.Vm.Vm.Value == "" {
		return errors.New("empty managed object reference")
	}

	return nil
}

func pdSendRequest(ctx context.Context, pdc *pdConfig, path string, pdObj pagerDutyData) ([]byte, error) {
	tpCfg := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: pdc.TlsInsecure,
		},
	}

	clt := &http.Client{
		Transport: tpCfg,
	}

	reqBody, err := json.Marshal(pdObj)
	if err != nil {
		return []byte{}, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", path, bytes.NewBuffer(reqBody))
	if err != nil {
		return []byte{}, fmt.Errorf("unable to create http request due to %w", err)
	}

	req.Header.Add("token", pdc.PagerDuty.RoutingKey)

	res, err := clt.Do(req)
	if err != nil {
		return []byte{}, fmt.Errorf("executing http request %w", err)
	}
	defer res.Body.Close()

	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return []byte{}, fmt.Errorf("reading http response body: %w", err)
	}

	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return []byte{}, fmt.Errorf("http response: %d, %+v", res.StatusCode, string(body))
	}

	return body, nil
}

File: examples/openfaas/python/trigger-pagerduty-incident/handler/handler.py
import sys, json, os
import urllib3
import requests
import traceback

# GLOBAL_VARS
DEBUG=False
class bgc:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

if(os.getenv("insecure_ssl")):
    # Surpress SSL warnings
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
if(os.getenv("write_debug")):
    sys.stderr.write(f"{bgc.WARNING}WARNING!! DEBUG has been enabled for this function. Sensitive information could be printed to sysout{bgc.ENDC} \n")
    DEBUG=True

def debug(s):
    if DEBUG:
        sys.stderr.write(s+" \n") #syserr only get logged on the console logs

#
### Paths and Endpoints
### More examples and details here - https://v2.developer.pagerduty.com/docs/send-an-event-events-api-v2
#
PAGERDUTY_API_PATH='https://events.pagerduty.com/v2/enqueue'
PD_CONFIG='/var/openfaas/secrets/pdconfig'
class FaaSResponse:
    """
    FaaSResponse is a helper class to construct a properly formatted message returned by this function.
    By default, OpenFaaS will marshal this response message as JSON.
    """    
    def __init__(self, status, message):
        """
        Arguments:
            status {str} -- the response status code
            message {str} -- the response message
        """        
        self.status=status
        self.message=message

class Pagerduty:
    """
    Pagerduty is a Class used to trigger Events in PagerDuty.
    """    

    def __init__(self,conn):
        """
        Arguments:
            conn {session} -- connection to PagerDuty REST API
        """
        self.session=conn

    # PagerDuty REST API implementation        
    def invoke(self,obj):
        """
        Make a rest api call to Pagerduty Events API
        
        Arguments:
            obj {dict} -- Generated API Body for the PagerDuty Events API

        Returns:
            FaaSResponse -- status code and message
        """
        try:
            resp = self.session.post(PAGERDUTY_API_PATH,json=obj)
            resp.raise_for_status()
            debug(f'{bgc.OKGREEN}HTTP POST Request successful{bgc.ENDC}')
            debug(f'{bgc.OKBLUE}Response Body > {bgc.ENDC}{resp.text}')
            resp_body = json.loads(resp.text)
            return FaaSResponse('200', 'Successfully invoked PagerDuty API! dedup_key for this request: {0}'.format(resp_body['dedup_key']))
        except requests.HTTPError as err:
            return FaaSResponse('500', 'Could not invoke PagerDuty API > HTTPError: {0}'.format(err))

def handle(req):
    
    # Validate Event input
    debug(f'{bgc.HEADER}---Validating CloudEvent--- {bgc.ENDC}')
    debug(f'{bgc.OKBLUE}Event (raw) > {bgc.ENDC}{req}')
    try:
        cevent = json.loads(req)
        debug(f'{bgc.OKGREEN}Successfully parsed Event into JSON!{bgc.ENDC}')
        debug(f'{bgc.OKBLUE}Event (JSON) > {bgc.ENDC}{json.dumps(cevent, indent=4, sort_keys=True)}')
    except json.JSONDecodeError as err:
        res = FaaSResponse('400','Invalid JSON > JSONDecodeError: {0}'.format(err))
        print(json.dumps(vars(res)))
        return
    
    # Validate Config file
    debug(f'{bgc.HEADER}---Validating Config--- {bgc.ENDC}')
    debug(f'{bgc.OKBLUE}Reading Config File > {bgc.ENDC}{PD_CONFIG}')
    try: 
        with open(PD_CONFIG, 'r') as pdconfigfile:
            pdconfig = json.load(pdconfigfile)
            debug(f'{bgc.OKGREEN}Successfully parsed Configuration into JSON!{bgc.ENDC}')
            debug(f'{bgc.OKBLUE}Configuration > {bgc.ENDC}{json.dumps(pdconfig,indent=4)}')
            routingkey=pdconfig['routing_key']
            event_action=pdconfig['event_action']
    except json.JSONDecodeError as err:
        res = FaaSResponse('400','Invalid JSON > JSONDecodeError: {0}'.format(err))
        print(json.dumps(vars(res)))
        return
    except KeyError as err:
        res = FaaSResponse('400','Required key not found in the provided configuration > KeyError: {0}'.format(err))
        print(json.dumps(vars(res)))
        return 
    except OSError as err:
        res = FaaSResponse('500','Could not read PagerDuty configuration > OSError: {0}'.format(err))
        print(json.dumps(vars(res)))
        return

    # Assert that the function is able to get the required information from the event and build the request body
    # For debugging: validate the JSON blob we received - uncomment print statements if needed
    # print(cevent)
    debug(f'{bgc.HEADER}---Building HTTP Request body--- {bgc.ENDC}')
    try:
        # Map the CloudEvent data and build the PagerDuty Event API Request body
        obj = {
                'routing_key': routingkey,
                'event_action': event_action,
                'client': 'VMware Event Broker Appliance',
                'client_url': cevent['source'],
                'payload': {
                    'summary': cevent['data']['FullFormattedMessage'],
                    'timestamp': cevent['data']['CreatedTime'],
                    'source': cevent['source'],
                    'severity': 'info',
                    'component': cevent['data']['Vm']['Name'],
                    'group': cevent['data']['Host']['Name'],
                    'class': cevent['subject'],
                    'custom_details': {
                        'user': cevent['data']['UserName'],
                        'Datacenter': cevent['data']['Datacenter'],
                        'ComputeResource': cevent['data']['ComputeResource'],
                        'Host': cevent['data']['Host'],
                        'VM': cevent['data']['Vm']
                    }
                }
                #,
                #'images': [{
                #    'src': 'https://www.pagerduty.com/wp-content/uploads/2016/05/pagerduty-logo-green.png',
                #    'href': 'https://example.com/',
                #    'alt': cevent['data']['Vm']
                # }],
                # 'links': [{
                #    'href': 'https://example.com/',
                #    'text': 'Link to VM'
                # }]
            }
        debug(f'{bgc.OKBLUE}Built API Request Body > {bgc.ENDC}{json.dumps(obj,indent=4)}')
    except KeyError as err:
        res = FaaSResponse('400','Invalid JSON, required key not found in the provided Event > KeyError: {0}'.format(err))
        print(json.dumps(vars(res)))
        traceback.print_exc(limit=1, file=sys.stderr) #providing traceback since it helps debug the exact key that failed
        return
    except TypeError as err:
        res = FaaSResponse('400','Invalid JSON, missing required data in the provided Event > TypeError: {0}'.format(err))
        print(json.dumps(vars(res)))
        traceback.print_exc(limit=1, file=sys.stderr) #providing traceback since it helps debug the exact key that failed
        return

    # Make the Rest Api Call to PagerDuty
    s=requests.Session()
    if(os.getenv("insecure_ssl")):
        s.verify=False
    debug(f'{bgc.HEADER}---Attemping API Request to PagerDuty--- {bgc.ENDC}')
    try:
        pg = Pagerduty(s)
        res = pg.invoke(obj)
        print(json.dumps(vars(res)))
    except Exception as err:
        res = FaaSResponse('500','Unexpected Error occurred > Exception: {0}'.format(err))
        print(json.dumps(vars(res)))
    
    #Close session
    s.close()

    return

#
## Unit Test - helps with executing the function locally
## Uncomment PDConfig - update the path to the file accordingly
## Uncomment handle('... to test the function with the event samples provided below test without deploying to OpenFaaS
#
#PD_CONFIG='pdconfig.json'

#
## FAILURE CASES :Invalid Inputs
#
#handle('')
#handle('"test":"ok"')
#handle('{"test":"ok"}')
#handle('{"data":"ok"}')

#
## FAILURE CASES :Unhandled Events
# 
# Standard : UserLogoutSessionEvent
#handle('{"id":"17e1027a-c865-4354-9c21-e8da3df4bff9","source":"https://vcsa.pdotk.local/sdk","specversion":"1.0","type":"com.vmware.event.router/event","subject":"UserLogoutSessionEvent","time":"2020-04-14T00:28:36.455112549Z","data":{"Key":7775,"ChainId":7775,"CreatedTime":"2020-04-14T00:28:35.221698Z","UserName":"machine-b8eb9a7f","Datacenter":null,"ComputeResource":null,"Host":null,"Vm":null,"Ds":null,"Net":null,"Dvs":null,"FullFormattedMessage":"User machine-b8ebe7eb9a7f@127.0.0.1 logged out (login time: Tuesday, 14 April, 2020 12:28:35 AM, number of API invocations: 34, user agent: pyvmomi Python/3.7.5 (Linux; 4.19.84-1.ph3; x86_64))","ChangeTag":"","IpAddress":"127.0.0.1","UserAgent":"pyvmomi Python/3.7.5 (Linux; 4.19.84-1.ph3; x86_64)","CallCount":34,"SessionId":"52edf160927","LoginTime":"2020-04-14T00:28:35.071817Z"},"datacontenttype":"application/json"}')
# Eventex : vim.event.ResourceExhaustionStatusChangedEvent
#handle('{"id":"0707d7e0-269f-42e7-ae1c-18458ecabf3d","source":"https://vcsa.pdotk.local/sdk","specversion":"1.0","type":"com.vmware.event.router/eventex","subject":"vim.event.ResourceExhaustionStatusChangedEvent","time":"2020-04-14T00:20:15.100325334Z","data":{"Key":7715,"ChainId":7715,"CreatedTime":"2020-04-14T00:20:13.76967Z","UserName":"machine-bb9a7f","Datacenter":null,"ComputeResource":null,"Host":null,"Vm":null,"Ds":null,"Net":null,"Dvs":null,"FullFormattedMessage":"vCenter Log File System Resource status changed from Yellow to Green on vcsa.pdotk.local  ","ChangeTag":"","EventTypeId":"vim.event.ResourceExhaustionStatusChangedEvent","Severity":"info","Message":"","Arguments":[{"Key":"resourceName","Value":"storage_util_filesystem_log"},{"Key":"oldStatus","Value":"yellow"},{"Key":"newStatus","Value":"green"},{"Key":"reason","Value":" "},{"Key":"nodeType","Value":"vcenter"},{"Key":"_sourcehost_","Value":"vcsa.pdotk.local"}],"ObjectId":"","ObjectType":"","ObjectName":"","Fault":null},"datacontenttype":"application/json"}')

#
## SUCCESS CASES
#
# Standard : VmPoweredOnEvent
#handle('{"id":"453120cd-3d19-4c43-aadc-df0cdbce3887","source":"https://vcsa.pdotk.local/sdk","specversion":"1.0","type":"com.vmware.event.router/event","subject":"VmPoweredOnEvent","time":"2020-04-13T23:46:10.402531287Z","data":{"Key":7441,"ChainId":7438,"CreatedTime":"2020-04-13T23:46:09.387283Z","UserName":"Administrator","Datacenter":{"Name":"PKLAB","Datacenter":{"Type":"Datacenter","Value":"datacenter-3"}},"ComputeResource":{"Name":"esxi01.pdotk.local","ComputeResource":{"Type":"ComputeResource","Value":"domain-s29"}},"Host":{"Name":"esxi01.pdotk.local","Host":{"Type":"HostSystem","Value":"host-31"}},"Vm":{"Name":"Test VM","Vm":{"Type":"VirtualMachine","Value":"vm-33"}},"Ds":null,"Net":null,"Dvs":null,"FullFormattedMessage":"Test VM on esxi01.pdotk.local in PKLAB has powered on","ChangeTag":"","Template":false},"datacontenttype":"application/json"}')
# Standard : VmPoweredOffEvent
#handle('{"id":"d77a3767-1727-49a3-ac33-ddbdef294150","source":"https://vcsa.pdotk.local/sdk","specversion":"1.0","type":"com.vmware.event.router/event","subject":"VmPoweredOffEvent","time":"2020-04-14T00:33:30.838669841Z","data":{"Key":7825,"ChainId":7821,"CreatedTime":"2020-04-14T00:33:30.252792Z","UserName":"Administrator","Datacenter":{"Name":"PKLAB","Datacenter":{"Type":"Datacenter","Value":"datacenter-3"}},"ComputeResource":{"Name":"esxi01.pdotk.local","ComputeResource":{"Type":"ComputeResource","Value":"domain-s29"}},"Host":{"Name":"esxi01.pdotk.local","Host":{"Type":"HostSystem","Value":"host-31"}},"Vm":{"Name":"Test VM","Vm":{"Type":"VirtualMachine","Value":"vm-33"}},"Ds":null,"Net":null,"Dvs":null,"FullFormattedMessage":"Test VM on  esxi01.pdotk.local in PKLAB is powered off","ChangeTag":"","Template":false},"datacontenttype":"application/json"}')
